### PR TITLE
all: support certificates in tls.external

### LINKED ISSUE #71: Support internal CA or self-signed TLS certificates for all external communication
- Issue type: Feature Request
- Sourcegraph version: 2.7.6
- OS Version: N/A
- Docker version: N/A

We have a private GitHub Enterprise instance running under a self-signed private Certificate Authority. Because of this, we need to tell the Sourcegraph image about this CA, or Git will complain:

```
fatal: unable to access 'https://github.example.com/user/repo.git': server certificate verification failed. 
```

The configuration option in [`github.certificate`](https://about.sourcegraph.com/docs/config/settings/#certificate-string) is available, but it only applies to Sourcegraph, not to processes that `gitserver` controls. As well, since it only seems to handle server certificates and not CAs, it will have to be updated every time the GitHub Enterprise server rotates its certs.

Our workaround, which you are welcome to document, is to install the certificate in the Docker image, so the OS handles it. In a `Dockerfile`:

```Dockerfile
FROM sourcegraph/server:2.7.6
COPY ssl/certificate-authority.crt /usr/local/share/ca-certificates
RUN /usr/sbin/update-ca-certificates
```

Running this derived image makes everything work magically, because Sourcegraph also trusts the CAs provided by the OS.

### PR DESCRIPTION
**Pull request by keegancsmith**
Extended the experimental tls.external site setting to include a list of
certificates. When set repo-updater will add the certificates to the HTTP
client's RootCAs. This features works on top of any certificates added to an
individual external service configuration. Additionally these certificates are
also set for remote git commands.

This was tested by running a local HTTPS proxy to GitHub.com. I ensured both
API requests and git requests only worked when the relevant tls.external
setting was configured.

Note: AWS CodeCommit still does not support these setting. However, I don't expect AWS CodeCommit to be self hosted so should be fine.

Fixes #71

### CODE REVIEW

**Comment by keegancsmith** [comment]
I did lots of manual testing. Will think about how to testing this in an automatic way (it can be modified to be go test, but is quite a lot of work for an unknown value). Here are my detailed notes on testing:
Generated a key and certificate

``` shell
openssl genrsa -out server.key 2048
openssl req -new -x509 -sha256 -key server.key -out server.crt -days 3650
```


Ran a proxy to github.com with it

``` go
package main

import (
    "log"
    "net/http"
    "net/http/httputil"
    "net/url"
    "strings"
)

func main() {
    u, err := url.Parse("https://api.github.com")
    if err != nil {
        log.Fatal(err)
    }
    urlDirector := httputil.NewSingleHostReverseProxy(u).Director
    director := func(req *http.Request) {
        orig := req.URL.String()
        urlDirector(req)
        if strings.HasPrefix(req.URL.Path, "/api/") {
            req.URL.Path = strings.TrimPrefix(req.URL.Path, "/api")
            req.URL.Path = strings.TrimPrefix(req.URL.Path, "/v3")
        } else {
            req.URL.Host = "github.com"
        }
        req.Host = req.URL.Host
        log.Println(orig, "->", req.URL)
    }
    err = http.ListenAndServeTLS(":443", "server.crt", "server.key", &httputil.ReverseProxy{
        Director: director,
    })
    if err != nil {
        log.Fatal("ListenAndServe: ", err)
    }
}
```

Then configured a GitHub external service pointing to my localhost. Everything
failed, but then tested with setting various `tls.external` settings:

``` json
"tls.external": {
    "insecureSkipVerify": true,
    "certificates": ["CERTHERE"]
}
```


Command line snippet to JSONify cert:

``` shell
echo '{}' | jq --arg val "$(cat test/server.crt)" '$val' | pbcopy
```

Additionally I had to modify the GitHub external service to use the baseURL
instead of the clone URL returned from the API

``` diff
diff --git a/cmd/repo-updater/repos/github.go b/cmd/repo-updater/repos/github.go
index fb5839966b..94d4284c84 100644
--- a/cmd/repo-updater/repos/github.go
+++ b/cmd/repo-updater/repos/github.go
@@ -305,6 +305,7 @@ func (s *GithubSource) authenticatedRemoteURL(repo *github.Repository) string {
        return repo.URL
    }
    u.User = url.User(s.config.Token)
+	u.Host = s.originalHostname
    return u.String()
 }
```

File: internal/httpcli/external.go [hunk 1]
```diff
@@ -45,6 +46,22 @@ func (t *externalTransport) update() *http.Transport {
 
 	effective.TLSClientConfig.InsecureSkipVerify = config.InsecureSkipVerify
 
+	for _, cert := range config.Certificates {
+		// There is no exposed Clone function for CertPools. So if a certificate
+		// is removed it will continue to be accepted since we are mutating base's
+		// RootCAs. This is an acceptable tradeoff since it would be quite tricky
+		// to avoid this.
```
Comment (unknwon) [inline internal/httpcli/external.go]: Would a restart of Sourcegraph after changing the "certificates" in site config result in a clean state (i.e. stop accepting removed certs)?

File: cmd/gitserver/server/serverutil.go [hunk 2]
```diff
@@ -129,6 +170,32 @@ func configureRemoteGitCommand(cmd *exec.Cmd, tlsConf *schema.TlsExternal) {
...
+		}
+	}()
+
+	n, err := f.Write(data)
+	if err == nil && n < len(data) {
+		err = io.ErrShortWrite
+	}
+	if err1 := f.Close(); err == nil {
+		err = err1
+	}
```
[the remark this answers is not in the document]
Comment (keegancsmith) [reply cmd/gitserver/server/serverutil.go]: this is copied from `ioutil.WriteFile`. I like your suggestion though, so will update.

File: internal/httpcli/external.go (hunk 1 above)
Comment (keegancsmith) [reply internal/httpcli/external.go]: yes

**Pull request merged**